java-viewer/src/main/java/org/sonar/java/viewer/Viewer.java

```diff
@@ -0,0 +1,129 @@
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import static spark.Spark.exception;
+import static spark.Spark.get;
+import static spark.Spark.port;
+import static spark.Spark.post;
+import static spark.Spark.staticFiles;
+
+import spark.ModelAndView;
+import spark.Request;
+import spark.template.velocity.VelocityTemplateEngine;
+
+public class Viewer {
+
+  private static final Logger LOGGER = Logger.getLogger(Viewer.class);
+  private static final String DEFAULT_SOURCE_CODE_LOCATION = "/public/example/example.java";
+  private static final String DEFAULT_SOURCE_CODE = defaultFileContent();
+
+  public static void main(String[] args) {
+    BasicConfigurator.configure();
+
+    exception(Exception.class, (e, req, res) -> LOGGER.error("Unexpected exception.", e)); // print all exceptions
+
+    staticFiles.location("/public");
+    port(9999);
+
+    get("/", (req, res) -> renderIndex(defaultFileContent()));
+    post("/", (req, res) -> generate(req));
+  }
+
+  private static String generate(Request request) {
+    MultiMap<String> params = new MultiMap<>();
+    UrlEncoded.decodeUtf8To(request.body(), params);
+    String javaCode = params.getOrDefault("javaCode", Collections.singletonList(DEFAULT_SOURCE_CODE)).get(0);
+    try {
+      return renderIndex(javaCode);
+    } catch (Exception e) {
+      return renderError(javaCode, e);
+    }
+  }
+
+  private static String renderIndex(String javaCode) {
+    HashMap<String, String> values = new HashMap<>();
+
+    CFG cfg = CFGViewer.buildCFG(javaCode);
+
+    values.put("cfg", CFGDebug.toString(cfg));
+
+    values.put("dotAST", ASTViewer.toDot(javaCode));
+    values.put("dotCFG", CFGViewer.toDot(cfg));
+    values.put("dotEG", EGViewer.toDot(javaCode, cfg.blocks().get(0).id()));
+
+    values.put("errorMessage", "");
+    values.put("errorStackTrace", "");
+
+    return renderWithValues(javaCode, values);
+  }
+
+  private static String renderError(String javaCode, Exception e) {
+    HashMap<String, String> values = new HashMap<>();
+
+    StringWriter sw = new StringWriter();
+    e.printStackTrace(new PrintWriter(sw));
+    String stackTrace = sw.toString();
+
+    String message = e.getMessage();
+    values.put("errorMessage", message == null ? "Unexpected error" : message);
+    values.put("errorStackTrace", stackTrace.replace(System.getProperty("line.separator"), "<br/>\n"));
+
+    return renderWithValues(javaCode, values);
+  }
+
+  private static String renderWithValues(String javaCode, Map<String, String> values) {
+    values.put("javaCode", javaCode);
+    return new VelocityTemplateEngine().render(new ModelAndView(values, "velocity/index.vm"));
+  }
+
+  private static String defaultFileContent() {
+    String result;
+    try {
+      Path path = Paths.get(Viewer.class.getResource(DEFAULT_SOURCE_CODE_LOCATION).toURI());
+      result = new String(Files.readAllBytes(path));
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this use of constructor "String(byte[])" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1943)